Initialize vptrs to point to vtables

dwblaikie opened the pull request:
Adds a mapping to keep track of vtable LLVM IR decls/defs for use.
Adds the vtable_id to the vtable_ptr initialize instruction for lookup.
Adds emission of vtable declarations for use outside the file that
defines the vtable. (this isn't done lazily, it's done for any imported
class - it could be done lazily & maybe eventually has to be lazy to
handle generics)

dwblaikie committed:
Initialize vptrs to point to vtables

Adds a mapping to keep track of vtable LLVM IR decls/defs for use.
Adds the vtable_id to the vtable_ptr initialize instruction for lookup.
Adds emission of vtable declarations for use outside the file that
defines the vtable. (this isn't done lazily, it's done for any imported
class - it could be done lazily & maybe eventually has to be lazy to
handle generics)

zygoloid commented on toolchain/lower/testdata/class/virtual.carbon:
```diff
@@ -106,6 +106,9 @@ fn Use() {
 // CHECK:STDOUT: ; ModuleID = 'create.carbon'
 // CHECK:STDOUT: source_filename = "create.carbon"
 // CHECK:STDOUT:
+// CHECK:STDOUT: @"_CIntermediate.Classes.$vtable" = external unnamed_addr constant ptr
+// CHECK:STDOUT: @"_CDerived.Classes.$vtable" = external unnamed_addr constant ptr
```
These are never used in the generated IR. It'd be useful to have a test that uses the intermediate or derived vtable, if that works, or a test with a TODO if not. What happens if we initialize the `b`, `i`, `d` variables in `create.carbon`?

zygoloid replied:
Mostly I'm asking because all the lowering test coverage here is assigning the vtable of the base-most class to the vptr, and it'd be nice to cover a case where we get a different vtable.

dwblaikie replied:
Ah, good call. (I probably should've split this patch in two and done the SemIR changes then the lower changes - to make it clear/work through this sort of thing)

Did turn out I was only ever using the base vtable, and it was visible in the existing test coverage - `lower/testdata/class/virtual.carbon` (oh, yes, the `create.carbon` you were mentioning) had three initializations all using the `Base` vtable, but should've been `Base`, `Base`, `Derived`.

I threaded vtable_id through the various conversions so now the most derived vtable is used to initialize the vptr inside the base subobject - but it's a bit awkward/probably not the best thing going through those layers. I'm open to suggestions. I guess this'll all change when we (probably me?) add partial class/object support, then the vptr field effectively won't exist until after the whole thing's been constructed and maybe it's easier/different to plumb it through then.

Alternatively I guess I could emit the vtable initialization as a separate instruction after the object initialization? Not sure if that's reasonable, though - again, open to ideas.

(also, vptrs aren't initialized when doing `var v: Derived;`, only when doing `var v: Derived = {...};` - I'll add that to my todo list)

Yeah, I don't have any support for or testing of virtual function calls - that's coming in a separate patch. Open to adding TODO testing for it here, or leaving it until later - I was planning to leave it to go with the patch that implements it & relying on SemIR/LLVMIR testing to show that the vtables exist, the vptrs are initialized correctly, etc.

zygoloid replied:
I think what you're doing here is fine for now, but as noted elsewhere we should have a TODO in the convert logic to remind ourselves it's not what we eventually want.

dwblaikie committed:
Correctly init vtable_ptr to point to the most derived vtable

This is a bit janky, weaving the vtable_id down through whatever layers
of conversion was needed. Open to other ideas, to be sure.

dwblaikie committed:
Add initializers to workaround missing vptr initialization

dwblaikie committed:
Fix typo

dwblaikie committed:
Test update

zygoloid commented on toolchain/check/convert.h:
```diff
@@ -56,7 +56,8 @@ struct ConversionTarget {
 auto Convert(Context& context, SemIR::LocId loc_id, SemIR::InstId expr_id,
-             ConversionTarget target) -> SemIR::InstId;
+             ConversionTarget target,
+             SemIR::InstId vtable_id = SemIR::InstId::None) -> SemIR::InstId;
```
Eventually, the vtable parameter here shouldn't be necessary (the vptr initialization should be handled by the initialization of a class of non-`partial` type, not the initialization of the base-most type). Please can you add a TODO somewhere to track this?

dwblaikie replied:
Yeah, for sure - added the TODO here on `Convert` (since that seems to be the place that this awkwardness is most exposed) in 27493b718c9e995c72ef7406aa0e0b8644b189b9

dwblaikie committed:
Add TODO comment regarding vtable_id init and eventual migration to handling this in partial-to-complete class type conversion
